### PR TITLE
[WIP] Allow function dispatch for constants

### LINKED ISSUE #1154: IncSubtensor causes graph break in pytorch backend
### Description

```py
import pytensor
import pytensor.tensor as pt
import torch

x = pt.vector('x')
y = x[-1].inc(1)
f = pytensor.function(inputs=[x], outputs=y, mode="PYTORCH")

torch._dynamo.explain(f.vm.jit_fn._fn)(torch.zeros(2)).break_reasons
```

WIll produce: `[GraphCompileReason(reason='data dependent operator: aten._local_scalar_dense.default; to enable, set torch._dynamo.config.capture_scalar_outputs = True', user_stack=[<FrameSummary file [/Users/ch0ronomato/dev/pytensor/pytensor/link/pytorch/dispatch/subtensor.py](https://file+.vscode-resource.vscode-cdn.net/Users/ch0ronomato/dev/pytensor/pytensor/link/pytorch/dispatch/subtensor.py), line 99 in inc_subtensor>], graph_break=True)]`

### PR DESCRIPTION
**Pull request by Ch0ronomato**
## Description
Pytorch doesn't seem to handle the constants we put in the globals of the generated code well. It can cause graph breaks for things that shouldn't be graph break (example: The index in IncSubtensor when the index is static). This change allows the typify method to instead return a function, and then defer fetching the constants in the generated code as a function call. 

A small example:
```py
import pytensor
import pytensor.tensor as ptt

x = ptt.vector('x')
o = x[-1].inc(1)
f = pytensor.function(inputs=[x], outputs=o, mode="PYTORCH")
```

will generate

```py
def pytorch_funcified_fgraph(x):
    tensor_constant = torch_assume_constant()
    scalar_constant = torch_assume_constant_1()
    # IncSubtensor{i}(x, 1, -1)
    tensor_variable = inc_subtensor(x, tensor_constant, scalar_constant)
    return (tensor_variable,)
```

`torch_assume_constant` can now do whatever it needs to do to allow the backend (torch in this case) to know that the variable is infact a constant. 

## Related Issue

- [ ] Closes #1154 
- [ ] Related to #

## Checklist

- [X] Checked that [the pre-commit linting/style checks pass](https://docs.pymc.io/en/latest/contributing/python_style.html)
- [ ] Included tests that prove the fix is effective or that the new feature works
- [ ] Added necessary documentation (docstrings and/or example notebooks)
- [X] If you are a pro: each commit corresponds to a [relevant logical change](https://wiki.openstack.org/wiki/GitCommitMessages#Structural_split_of_changes)

## Type of change

- [X] New feature / enhancement
- [X] Bug fix
- [ ] Documentation
- [ ] Maintenance
- [ ] Other (please specify):

----
📚 Documentation preview 📚: https://pytensor--1159.org.readthedocs.build/en/1159/

### CODE REVIEW

File: pytensor/link/pytorch/dispatch/basic.py [hunk 1]
```diff
@@ -36,11 +36,15 @@ def pytorch_typify_tensor(data, dtype=None, **kwargs):
...
 @pytorch_typify.register(slice)
 @pytorch_typify.register(NoneType)
-@pytorch_typify.register(np.number)
 def pytorch_typify_no_conversion_needed(data, **kwargs):
     return data
 
 
+@pytorch_typify.register(np.number)
+def pytorch_typify_extract(data, **kwargs):
+    return data.item()
```
Comment (ricardoV94) [inline pytensor/link/pytorch/dispatch/basic.py:45]: This seems wrong?

**Review by ricardoV94 — COMMENTED** [review]
So did it work?

File: pytensor/link/pytorch/dispatch/basic.py (hunk 1 above)
Comment (Ch0ronomato) [reply pytensor/link/pytorch/dispatch/basic.py:45]: Say more? 

The torch compiler threw asserts when a zero dim np value was passed back.

**Comment by Ch0ronomato** [comment]
It worked for the very simple case. I'm gonna try it with the 8 schools model and see if that works. 

It also broke a few tests lol

File: pytensor/link/pytorch/dispatch/basic.py (hunk 1 above)
Comment (ricardoV94) [reply pytensor/link/pytorch/dispatch/basic.py:45]: Isn't it going to upcast values to float64, whatever python integers are? Does torch have scalars (not tensors) with specific dtypes we can use instead?

**Comment by ricardoV94** [comment]
Interestingly enough we have been considered defining constants as zero input Ops to make code more clean in rewrites (we always have to worry about whether a variable has an owner, which constants and root inputs do not).

**Comment by Ch0ronomato** [comment]
> Interestingly enough we have been considered defining constants as zero input Ops to make code more clean in rewrites (we always have to worry about whether a variable has an owner, which constants and root inputs do not).

That makes a lot of sense to me. I would be happy to sketch that out.

File: pytensor/link/pytorch/dispatch/basic.py [hunk 2]
```diff
@@ -57,11 +61,20 @@ def pytorch_funcify_FunctionGraph(
     conversion_func=pytorch_funcify,
     **kwargs,
 ):
+    def constants_wrapper(x, **kwargs):
+        x = pytorch_typify(x)
+
+        @torch.compiler.assume_constant_result
+        def torch_assume_constant(arg=x):
```
Comment (Ch0ronomato) [inline pytensor/link/pytorch/dispatch/basic.py:68]: this unfortunately suffers from the same problem some of the generated functions were getting where pytorch loses their reference somewhere in the process. We'll need to make sure it gets added to the `warpper::gen_functors` list

File: pytensor/link/utils.py [hunk 3]
```diff
@@ -749,9 +749,23 @@ def fgraph_to_python(
...
+                if callable(getter_or_value):
+                    # we got passed a function, this could be used to indicate something
+                    # to the backend. We'll embed it
+                    new_output_name = unique_name(i)
+                    getter_unique_name = unique_name(getter_or_value)
+                    global_env[getter_unique_name] = getter_or_value
+                    assign_str = f"{new_output_name} = {getter_unique_name}()"
+                    body_assigns.append(assign_str)
+                    node_input_names.append(new_output_name)
+                    continue
```
Comment (Ch0ronomato) [inline pytensor/link/utils.py:764]: Hopefully we don't need this and it was an over optimization. Try to refactor away the continue.

File: pytensor/link/pytorch/linker.py [hunk 4]
```diff
@@ -25,9 +27,21 @@ def conversion_func_register(*args, **kwargs):
             self.gen_functors.append((f"_{name}", functor))
             return functor
 
+        def constants_wrapper(x, **kwargs):
+            x = pytorch_typify(x)
+
+            @torch.compiler.assume_constant_result
+            def torch_assume_constant(arg=x):
```
Comment (Ch0ronomato) [inline pytensor/link/pytorch/linker.py:34]: This is causing some strange behavior in the compiler where `outs` is a bunch of functions (the constants specifically) which doesn't make a lot of sense to me. I'm still investigating the cause.

Comment (Ch0ronomato) [reply pytensor/link/pytorch/linker.py:34]: It's possible this approach has issues since we don't wrap things in a nn.Module https://github.com/pytorch/pytorch/issues/124858 (we shouldn't need to.)

**Comment by Ch0ronomato** [comment]
I'm gonna close this out. It's not obvious that we get any benefit from it since there are a lot of footguns to get it working
